File: py_app/util/wallet/okx.py
from time import sleep
from DrissionPage import WebPage
from loguru import logger
from py_app.util.browser.browser import click_by, input_content, wait_for
from py_app.util.wallet.metamask import WalletDoSomethingOption

# Constants
OKX_EXTENSION_ID = "mcohilncbfahbmgdjkbpemcciiolgcge"
OKX_PASSWORD = "abc123456"
DEFAULT_WAITE_TIME=5
DEFAULT_SLEEP_TIME = 2
MAX_RETRIES = 10
MAX_IDLE_ATTEMPTS = 10


def ImportOKXByPrivateKey(browser: WebPage, private_key: str) -> bool:
    """Import an account into OKX wallet using private key."""
    if not private_key or len(private_key) < 40:
        logger.error("ImportOKXByPrivateKey() received invalid private_key")
        return False

    if not browser:
        logger.error("ImportOKXByPrivateKey() received invalid browser")
        return False

    try:
        sleep(DEFAULT_WAITE_TIME)
        page = browser.new_tab(f"chrome-extension://{OKX_EXTENSION_ID}/notification.html")
        page.clear_cache()
        page.set.activate()
        wait_for(page)
        sleep(DEFAULT_SLEEP_TIME)

        # Check if password unlock is needed
        if "请输入密码" in page.html:
            if not input_content(page=page, ele="@type=password", content=OKX_PASSWORD):
                return False
            wait_for(page)
            if not click_by(page=page, ele="tag:button@text()=解锁"):
                return False
            wait_for(page)
            page.close()
            return True

        # Execute import steps
        if not _execute_import_steps(page, private_key):
            page.close()
            return False

        page.close()
        return True

    except Exception as e:
        logger.error(f"OKX import failed: {e}")
        if 'page' in locals():
            page.close()
        return False


def _execute_import_steps(page: WebPage, private_key: str) -> bool:
    """Execute the step-by-step import process."""
    steps = [
        # (description, action_type, selector, content, timeout)
        ("Find import wallet option", "click", 'tag:button@|text():导入已有钱包@|text():Import wallet', None, 60),
        ("Select private key option", "click", 'tag:div@|text()=Seed phrase or private key@|text()=助记词或私钥', None, 30),
        ("Choose private key method", "click", 'x://div[@data-e2e-okd-tabs-pane="2"]', None, None), #
        ("Input private key field", "input", '@type=password', private_key, None),
        ("Confirm private key", "click", 'x:(//button[@data-testid="okd-button"][1])', None, None),
        ("Second confirmation", "click", 'x:(//button[@data-testid="okd-button"][2])', None, None),
        ("Password verification", "click", 'tag:div@|text():密码验证@|text()=Password', None, None),
        ("Next step button", "click", 'x://button[@data-testid="okd-button"]', None, None),
        ("First password input", "input", "x:(//input[@type='password'])[1]", OKX_PASSWORD, None),
        ("Confirm password input", "input", "x:(//input[@type='password'])[2]", OKX_PASSWORD, None),
        ("Final confirmation", "click", 'x://button[@data-testid="okd-button"]', None, None),
        ("Complete onboarding", "click", 'x://button[@data-testid="okd-button"]', None, 2)
    ]

    for step in steps:
        desc, action, selector, content, timeout = step
        logger.debug(f"Executing step: {desc}")

        try:
            wait_for(page, ele=selector, timeout=timeout or 30)
            # sleep(DEFAULT_SLEEP_TIME)

            if action == "click":
                if not click_by(page=page, ele=selector, timeout=timeout):
                    logger.error(f"Failed at step: {desc}")
                    return False
            elif action == "input":
                if not input_content(page=page, ele=selector, content=content):
                    logger.error(f"Failed at step: {desc}")
                    return False

            sleep(0.5)
        except Exception as e:
            logger.error(f"Error at step '{desc}': {str(e)}")
            return False

    return True


def ClickBtnPrimary(page: WebPage) -> bool:
    """Click the primary button in OKX interface."""
    return click_by(page, "@|class:btn-fill-highlight", timeout=5)


def OkxDoSomething(browser: WebPage, opt=None) -> bool:
    """Perform operations with OKX wallet."""
    if not browser:
        logger.error("OkxDoSomething() received invalid browser")
        sleep(5)
        return False

    options = opt or WalletDoSomethingOption()
    idle_count = 0

    for _ in range(MAX_RETRIES):
        if idle_count >= options.MaxIdle:
            break

        sleep(1)
        try:
            wallet_page = _get_okx_tab(browser)
            if not wallet_page:
                idle_count += 1
                sleep(3)
                continue

            sleep(0.5)
            if _check_insufficient_balance(wallet_page):
                return False

            for _ in range(MAX_RETRIES):
                if not ClickBtnPrimary(wallet_page):
                    break

        except Exception:
            idle_count += 1
            sleep(3)
            continue

    _close_okx_tab(browser)
    return True


def _get_okx_tab(browser: WebPage):
    """Get the OKX extension tab if it exists."""
    try:
        return browser.get_tab(url=OKX_EXTENSION_ID)
    except Exception:
        return None


def _close_okx_tab(browser: WebPage) -> None:
    """Close the OKX extension tab if it exists."""
    try:
        if tab := browser.get_tab(url=OKX_EXTENSION_ID):
            tab.close()
    except Exception:
        pass


def _check_insufficient_balance(page: WebPage) -> bool:
    """Check for insufficient balance warning."""
    insufficient = page.ele('@|text():足够@|text():enough', timeout=5)
    if insufficient:
        logger.error("Insufficient balance, aborting operation")
        click_by(page, 't:button@data-testid=page-container-footer-cancel')
        return True
    return False

File: ts_app/utils/wallet/okx.ts
import { Browser, Page } from "rebrowser-puppeteer-core";
import { sleep } from "../help";
import {
  click,
  clickIfAvailable,
  disableNewPages,
  executeSteps,
  findPage,
  has,
  hs,
  input,
  newPage,
  Step,
  wait,
  wrapSelector,
} from "../browser/page";
import logger from "../../infrastructure/logger";
import { InsufficientBalance } from "../../types/error";

// 常量
const OKX_EXTENSION_ID = "mcohilncbfahbmgdjkbpemcciiolgcge";
const OKX_EXTENSION_URL = `chrome-extension://${OKX_EXTENSION_ID}/notification.html`;
const OKX_PASSWORD = "abc123456";
const DEFAULT_WAIT_TIME = 5000;
const DEFAULT_SLEEP_TIME = 2000;

async function getWalletPage(browser: Browser): Promise<Page | null> {
  try {
    const walletPage = await findPage(browser, {
      urlContain: OKX_EXTENSION_ID,
    });
    if (walletPage && walletPage.ok) return walletPage.data;

    const page = await newPage(browser, OKX_EXTENSION_URL);
    return page;
  } catch (error) {
    logger.warn("Retrying wallet page...");
    await sleep(1000);
    return await newPage(browser, OKX_EXTENSION_URL);
  }
}

export async function importOKXByPrivateKey(
  browser: Browser,
  privateKey: string,
): Promise<boolean> {
  if (!privateKey || privateKey.length < 40) {
    console.error("importOKXByPrivateKey() received invalid private_key");
    return false;
  }
  const page = await getWalletPage(browser);
  await sleep(DEFAULT_SLEEP_TIME);

  const restore = disableNewPages(browser, page);
  try {
    await sleep(DEFAULT_WAIT_TIME);

    if (await has(page, wrapSelector("//*[contains(text(),'请输入密码')]"))) {
      await input(page, "input[type='password']", OKX_PASSWORD);
      await click(page, "//button[text()='解锁']");
      await sleep(DEFAULT_SLEEP_TIME);
      await page.close();
      return true;
    }

    // 执行导入步骤
    const success = await executeImportSteps(page, privateKey);
    await page.close();
    return success;
  } catch (e) {
    console.error(`OKX import failed: ${e}`);
    return false;
  } finally {
    restore();
  }
}

async function executeImportSteps(
  page: Page,
  privateKey: string,
): Promise<boolean> {
  // 导入步骤
  const steps: Step[] = [
    hs(
      "click",
      `${wrapSelector('//span[text()="导入已有钱包"]')}, ${wrapSelector("//span[text()='Import wallet']")}`,
      800,
    ),
    hs(
      "click",
      `${wrapSelector("//div[text()='Seed phrase or private key']")}, ${wrapSelector("//div[text()='助记词或私钥']")}`,
      800,
    ),
    hs("click", "div[data-e2e-okd-tabs-pane='2']", 800),
    hs("input", 'textarea[data-testid="okd-input"]', 800, privateKey),
    hs("click", "(//button[@data-testid='okd-button'])[1]", 800),
    hs("click", "(//button[@data-testid='okd-button'])[1]", 800),
    hs(
      "click",
      `${wrapSelector("//div[text()='密码验证']")}, ${wrapSelector("//div[text()='Password']")}`,
      800,
    ),
    hs("click", "button[data-testid='okd-button']", 800),
    hs("input", "(//input[@type='password'])[1]", 800, OKX_PASSWORD),
    hs("input", "(//input[@type='password'])[2]", 800, OKX_PASSWORD),
    hs("click", "button[data-testid='okd-button']", 800),
    hs("click", "button[data-testid='okd-button']", 2000),
  ];

  await executeSteps(page, steps);

  return true;
}

export const doSomething = async (
  browser: Browser,
  idleTimes: number = 5,
  must: boolean = false,
): Promise<boolean> => {
  let called = false;
  let idle = 0;
  let okTimes = 0;
  let notFoundPrimaryBtn = 0;

  for (let i = 0; i < 60; i++) {
    await sleep(1000);

    if (notFoundPrimaryBtn > 5) {
      throw new Error(
        `钱包插件连续 ${notFoundPrimaryBtn} 次未找到可点击按钮，可能是余额不足或网络错误`,
      );
    }

    const walletPage = await findPage(browser, {
      urlContain: OKX_EXTENSION_ID,
    });
    if (!walletPage || !walletPage.ok) {
      // logger.warn(`未找到插件页`);

      if (must && okTimes < 1) {
        i = 0; // 强制重试
        continue;
      }
      if (idle >= idleTimes) {
        logger.info(`钱包处理逻辑完成，退出循环`);
        break;
      }
      idle++;
      continue;
    }

    const pluginPage = walletPage.data;

    await pluginPage.bringToFront();
    called = true;
    idle = 0; // 一旦找到了插件页并处理，重置 idle

    logger.info(`检查插件按钮状态，当前页面URL: ${pluginPage.url()}`);

    try {
      if (await checkInsufficientBalance(pluginPage)) {
        logger.error(`当前交易 余额不足`);
        throw new InsufficientBalance("当前交易 余额不足");
      }

      await wait(pluginPage, ".btn-fill-highlight");
      console.log("start something");
      const success = await clickIfAvailable(pluginPage, `.btn-fill-highlight`);

      if (success) {
        okTimes++;
        notFoundPrimaryBtn = 0;
        logger.info(`成功点击插件按钮，累计成功次数: ${okTimes}`);
      } else {
        notFoundPrimaryBtn++;
        logger.warn(
          `找到按钮但不可点击 (disabled)，计数 +1 -> ${notFoundPrimaryBtn}`,
        );
      }
    } catch (e: any) {
      notFoundPrimaryBtn++;
      logger.warn(`点击插件按钮失败: ${e.message}`);
    }

    await sleep(500);
  }

  return called;
};

async function checkInsufficientBalance(page: Page): Promise<boolean> {
  const nodes = await page.$$(
    wrapSelector("//*[contains(text(),'足够') or contains(text(),'enough')]"),
  );
  if (nodes.length > 0) {
    console.error("Insufficient balance");
    await click(page, "button[data-testid='page-container-footer-cancel']");
    return true;
  }
  return false;
}
